### ISSUE TITLE
[Feature] Allow server commands to be called

### ISSUE THREAD

**Issue by andrewrkerr** [appsmithorg/appsmith#2404]
## Summary

Allowing server commands (bash, powershell, etc.) to be called and return output.

## Motivation

This would enable automation that isn't handled easily by an API/DB. Basic example: IT creates a basic form/button that would call a command to disable someone in AD. HR could just put in a username a click disable. This could then send an email to a distro letting them know the user has been disabled. This is just an example, but the possibilities for scripting would be endless.

## Additional Context

Using JS would run the command locally, but the user may not be able to run the command due to vlan, permissions, or other factors. Having the command run from the server could resolve this. IT would be able to handle the logic, error checking and etc within their code. The main idea is just to be able to call commands from Appsmith to the server.

Security could be a concern since you may not want all users to have access to this feature. Permissions around it would be a bonus.

**Comment by sumitsum** [comment]
Hi @andrewrkerr, can you please explain if it is same or different from running an REST API call on the server ?

**Comment by andrewrkerr** [comment]
No, it wouldn't be an api call. Here's a better breakdown of what I meant.

I install Appsmith on a linux server. I'd like to be able to send a bash command from Appsmith to the server it's installed on. Another example: I set a button so that when someone clicks it, then it can run "apt-get update" or any other bash command. 

Let me know if this makes more sense.

**Comment by sumitsum** [comment]
Hi @andrewrkerr, thanks for clearing that up. Can you please explain a little more regarding the use case and what more can be done with it ?

**Comment by andrewrkerr** [comment]
The use case would be to allow IT to create automations and allow for the "non-tech" to run the commands with a button click. Here's a few examples:

IT installs powershell on the same server that Appsmith is installed. IT creates powershell scripts for various tasks. IT links the powershell scripts to different buttons/forms in Appsmith. Appsmith calls the powershell command/script and gives the "non-tech" access to the page with the buttons/forms/etc. This would allow us to automate almost anything that isn't an API or database call.

Most of our users don't have access to do things like: disable users in AD, purge emails from O365, install apps to other servers., and many other tasks. This would allow IT to provide a simple option to the end users to handle these tasks and more. The automation possibilities with this type of setup is almost endless.

**Comment by sumitsum** [comment]
Hi @Nikhil-Nandagopal @mohanarpit need your opinion on this one. Do you think the use case described fit the Appsmith product vision/goal ?

**Comment by andrewrkerr** [comment]
Hey @sumitsum - For #1 - I understand if it doesn't align with your vision/goal. For #2 - I think this could be handled with permissions to alleviate the security concerns. 

Either way, I think it's a great tool that will help lots of people.

**Comment by mohanarpit** [comment]
@andrewrkerr If I re-frame this feature request, do you think your use-case can be accomplished using an SSH plugin? Imagine a case where you could provide your SSH username, password/identity file to Appsmith and setup a SSH connection. You could then SSH into any server you'd like to and execute any shell command on the remote server. Technically, you could also SSH into the same server where Appsmith is hosted and run any shell commands.

This would be secure because the permissions for the commands being executed will depend on the securities of your SSH login user.

As @sumitsum mentioned, I don't think allowing an application to modify the underlying OS is a good idea. These permissions should be granted explicitly.

**Comment by Nikhil-Nandagopal** [comment]
@andrewrkerr thank you for the details. This is a pretty unique idea! It sounds like an alternative to having to write an API / DB query for users who are more comfortable using a scripting language.

**Comment by andrewrkerr** [comment]
@mohanarpit That would work as well!

**Comment by rubabuddin** [comment]
Would also be interested in something similar so that Python scripts that live on another server could be executed. This would remove the requirement of exposing these as APIs.

**Comment by richardkaplan** [comment]
A similar example would be the abilitly to run an Applescript or Javascript for Automation script - you would never want to expose this to random people, but for use by trusted people it could be extremely useful.

**Comment by TannerW** [comment]
Any update on this? Really exciting feature!

**Comment by Nikhil-Nandagopal** [comment]
Unfortunately not. We have a backlog of plugins that we haven't been able to get to lately. We're working on creating bandwidth to pick these up!

**Comment by TannerW** [comment]
@Nikhil-Nandagopal Understood! Thank you for the response and the hard work!

**Comment by CanalWood** [comment]
I've seen the video of this new feature and it's really exciting!

I would like to use this feature to run script to execute some powercli vsphere command or execute terraform file.  
Do you have an estimate date for this delivery?

Good luck and thanks!!

**Comment by sumitsum** [comment]
@CanalWood thanks for sharing your feedback and use case :) . Regarding the timeline, this quarter seems difficult, next quarter may be possible, not sure though. Apologies for the delay.

**Comment by Lars-Br** [comment]
Has this feature somehow been canceled? We are eagerly waiting for it, but there have been no updates in quite some time, and the feature-branch that already contained a working plugin has been deleted. At least this feature request is still open, so is there still hope this might be implemented?

**Comment by rohan-arthur** [comment]
@Lars-Br we haven't been able to prioritise this feature. Meanwhile we're working on making the js objects more stable. 
Would you please share the scenarios you're looking to realise with appsmith where the SSH integration would help you?

**Comment by pgampe** [comment]
@rohan-arthur Shell scripting has been the around for 50 years. Most of the IT is automated with this. Having the option to run remote SSH commands would connect Appsmith with the power of 50 years of IT. There are battle tested security solutions around scripts and SSH remote commands.

Most admin scripts are not exposed on the web for good reasons (security, maintainability). 

The most specific use cases I have in mind:
- Copy a database from prod to dev (via shells script)
- Sync files between servers (via shell script)
- Trigger backup/ recovery jobs
- Trigger service restarts
- Trigger certain actions on servers where a local admin tool or scripts already exist

It would also allow Appsmith to service as a simple frontend for admin scripts that would otherwise be hard to expose on the web (because of security, maintenance, etc).

The way it is currently done is that you would need to develop a local webserver around the scripts to be able to do this with Appsmith. But then you would not need Appsmith in the first place.

**Comment by carinanfonseca** [comment]
Thanks for your input, @pgampe! I’m currently evaluating this and would love to discuss it with you. Feel free to book a time that works for you [here](https://calendly.com/carina-neves-fonseca/appsmith?month=2025-02).
